src/Task.cc

```diff
@@ -3694,6 +3694,73 @@ static void copy_mem_mapping(Task* from, Task* to, const KernelMapping& km) {
   }
 }
 
+static bool copy_mem_mapping_just_used(Task* from, Task* to, const KernelMapping& km)
+{
+  ScopedFd& fd = from->pagemap_fd();
+  if (!fd.is_open()) {
+    LOG(debug) << "Failed to open " << from->proc_pagemap_path();
+    return false;
+  }
+
+  size_t pagesize = page_size();
+
+  vector<uint64_t> buf;
+#if defined(__i386__)
+  const auto buf_page_count = 64*1024;
+#else
+  const auto buf_page_count = 1024*1024;
+#endif
+  buf.resize(buf_page_count);
+
+  uint64_t pages_present = 0; // Just for logging
+
+  for (uintptr_t offset = 0; offset < km.size(); offset += buf.size() * pagesize) {
+
+    auto read_offset = ((km.start().as_int() + offset) / pagesize) * sizeof(uint64_t);
+    auto read_count = buf.size() * sizeof(uint64_t);
+    ssize_t bytes_read = pread(fd, buf.data(), read_count, read_offset);
```

Comment:
I think we should limit the read to min(number of pages left in the mapping, max buffer size).
Also, we should be able to limit the buffer size allocation to the number of pages we need to read by calling resize() inside this loop. Then we don't allocate an over-large buffer.